src/replication.c

```diff
@@ -2622,6 +3261,131 @@ int slaveTryPartialResynchronization(connection *conn, int read_reply) {
+ * ┌────────▼──────────┐       │   ┌─▼─────────────────┐ │  │  ┌───────▼───────────────────────┐ │  ┌──▼────────────────┐  │    
+ * │SEND_HANDSHAKE     │       │ ┌─┤RECEIVE_PSYNC_REPLY├─┘  │  │RDB_CONN_RECEIVE_AUTH_REPLY    │ │  │RECEIVE_CAPA_REPLY │  │    
+ * └────────┬──────────┘       │ │ └─┬─────────────────┘    │  └───────┬───────────────────────┘ │  └──┬────────────────┘  │    
+ *          │                  │ │   │+FULLRESYNC           │          │+OK                      │     │+OK                │    
+ * ┌────────▼──────────┐       │ │ ┌─▼─────────────────┐    │  ┌───────▼───────────────────────┐ │  ┌──▼────────────────┐  │    
+ * │RECEIVE_AUTH_REPLY │       │ │ │TRANSFER           │    │  │RDB_CONN_RECEIVE_REPLCONF_REPLY│ │  │SEND_PSYNC         │  │    
+ * └────────┬──────────┘       │ │ └───────────────────┘    │  └───────┬───────────────────────┘ │  └──┬────────────────┘  │    
+ *          │+OK               │ │                          │          │+OK                      │     │PSYNC use snapshot │    
+ * ┌────────▼──────────┐       │ │                          │  ┌───────▼───────────────┐         │     │end-offset provided│    
+ * │RECEIVE_PORT_REPLY │       │ │                          │  │RDB_CONN_RECEIVE_ENDOFF│         │     │by the master      │    
+ * └────────┬──────────┘       │ │                          │  └───────┬───────────────┘         │  ┌──▼────────────────┐  │    
+ *          │+OK               │ │+CONTINUE                 │          │$ENDOFF                  │  │RECEIVE_PSYNC_REPLY│  │    
+ * ┌────────▼──────────┐       │ │                          │          ├─────────────────────────┘  └──┬────────────────┘  │    
+ * │RECEIVE_IP_REPLY   │       │ │                          │          │                               │+CONTINUE          │    
+ * └────────┬──────────┘       │ │                          │  ┌───────▼───────────────┐            ┌──▼────────────────┐  │    
+ *          │+OK               │ │                          │  │RDB_CONN_RDB_LOAD      │            │TRANSFER           │  │    
+ * ┌────────▼──────────┐       │ │                          │  └───────┬───────────────┘            └────┬──────────────┘  │    
+ * │RECEIVE_IP_REPLY   │       │ │                          │          │Done loading snapshot            │                 │    
+ * └────────┬──────────┘       │ │                          │  ┌───────▼───────────────┐                 │                 │    
+ *          │+OK               │ │                          │  │RDB_CONN_RDB_LOADED    │                 │                 │    
+ * ┌────────▼────────────────┐ │ │                          │  └───────┬───────────────┘                 │                 │    
+ * │RECEIVE_NO_FULLSYNC_REPLY│ │ │                          │          │                                 │                 │    
+ * └─┬────┬──────────────────┘ │ └────┐                     │          │Slave loads local replication    │                 │    
```

Comment:
Hi Viktor, thank you for your comment!
I agree with your suggested flow change, postponing the check of master capabilities makes things clearer on the replica side.

Regarding the diagram style, I used [asciiflow](https://asciiflow.com/), so I'll apply your fixes with some modifications to ensure we can continue using this tool for future changes.

As for point 1, the joined state where the main replication state is `TRANSFER` and the RDB connection state is `RDB_CONN_RDB_LOAD` indicates a currently full-syncing replica using the RDB channel. We could consider reusing `CONNECTED` after a `+CONTINUE`, but it doesn't make sense to me because the replica is currently "partial syncing" and isn't actually connected.